File: app/src/main/java/com/rorpage/purtyweather/util/WeatherUpdateScheduler.java
package com.rorpage.purtyweather.util;

import android.app.job.JobInfo;
import android.app.job.JobScheduler;
import android.content.ComponentName;
import android.content.Context;

import com.rorpage.purtyweather.services.PurtyWeatherService;

public class WeatherUpdateScheduler {
    public static void scheduleJob(Context context, int minimumLatencyMinutes) {
        ComponentName serviceComponent = new ComponentName(context, PurtyWeatherService.class);

        JobInfo.Builder builder = new JobInfo.Builder(0, serviceComponent);
        builder.setMinimumLatency(minimumLatencyMinutes * 60000);

        JobScheduler jobScheduler = context.getSystemService(JobScheduler.class);
        jobScheduler.schedule(builder.build());
    }

    public static void scheduleJob(Context context) {
        scheduleJob(context, 15);
    }
}

File: app/src/main/java/com/rorpage/purtyweather/util/WeatherUpdateScheduler.kt
package com.rorpage.purtyweather.util

import android.content.Context
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.OutOfQuotaPolicy
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import com.google.accompanist.pager.ExperimentalPagerApi
import com.rorpage.purtyweather.services.BackgroundWork
import java.util.concurrent.TimeUnit

object WeatherUpdateScheduler {

    val tagName = "weatherApiCall"

    @ExperimentalPagerApi
    @JvmStatic
    @JvmOverloads
    fun scheduleJob(context: Context, fromUI: Boolean) {
        val minimumLatencyMinutes = 15L

        val wm = WorkManager.getInstance(context)

        val periodicWorkRequest = PeriodicWorkRequestBuilder<BackgroundWork>(minimumLatencyMinutes, TimeUnit.MINUTES)
            .addTag(tagName)

        if (fromUI) {
            val workRequest = OneTimeWorkRequestBuilder<BackgroundWork>()
                .addTag(tagName)
                .setExpedited(OutOfQuotaPolicy.RUN_AS_NON_EXPEDITED_WORK_REQUEST)
                .build()
            wm.enqueue(workRequest)

            periodicWorkRequest.setInitialDelay(minimumLatencyMinutes, TimeUnit.MINUTES)
        }

        // even if we're from the UI, make sure this is scheduled.
        wm.enqueueUniquePeriodicWork(tagName, ExistingPeriodicWorkPolicy.REPLACE, periodicWorkRequest.build())
    }
}
